src/measure_theory/function/intersectivity.lean

```diff
@@ -0,0 +1,122 @@
+/-
+Copyright (c) 2023 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import measure_theory.function.lp_space
+import measure_theory.integral.average
+import order.upper_lower.locally_finite
+
+/-!
+# Bergelson's intersectivity lemma
+
+This file proves the Bergelson intersectivity lemma: In a finite measure space, a sequence of events
+that have measure at least `r` has an infinite subset whose finite intersections all have positive
+volume.
+
+This is in some sort a finitary version of the second Borel-Cantelli lemma.
+
+## References
+
+[Bergelson, *Sets of recurrence of `ℤᵐ`-actions and properties of sets of differences in
+`ℤᵐ`][bergelson1985]
+
+## TODO
+
+Restate the theorem using the upper density of a set of naturals, once we have it.
+
+Use the ergodic theorem to deduce the refinement of the Poincaré recurrence theorem proved by
+Bergelson.
+-/
+
+open filter measure_theory set
+open_locale big_operators ennreal nnreal
+
+variables {α : Type*} [measurable_space α] {μ : measure α} [is_finite_measure μ] {s : ℕ → set α}
```

Comment:
Could you please replace `nat` with an arbitrary infinite set? If the proof (didn't read yet) relies on `nat`, then you can have an auxiliary statement about `nat`, then apply it to `s` composed with an embedding of `nat` into your index type.

Reply:
The strong statement needs `nat` (it talks about natural density), but indeed the current statement can be generalised to any type (although I suspect that's mathematically irrelevant).